### ISSUE TITLE
RunDirectory does not work as context manager when voview file present

### ISSUE THREAD

**Issue by cydanil** [European-XFEL/EXtra-data#374]
[In the Jungfrau correction notebook, the raw run files are opened using a context manager](https://git.xfel.eu/detectors/pycalibration/-/blob/master/notebooks/Jungfrau/Jungfrau_Gain_Correct_and_Verify_NBC.ipynb#L697).  

This works fine when no voview files were created, as the files are opened directly.  
It's also fine to use the voview file when using `RunDirectory` as a regular object.  
However, when opening a run using a voview file ([the default behaviour when available](https://github.com/European-XFEL/EXtra-data/blob/d4c547c3af9a993ec2da41b5bf3d830dce9ab43e/extra_data/reader.py#L1398:L1401)), an [error is raised](https://github.com/European-XFEL/EXtra-data/blob/master/extra_data/reader.py#L186:L194) as files cannot be closed when leaving the context.  

We noticed this recently as HED is using the RunValidator which creates such files, whereas other instruments do not (yet).

## How to Reproduce  

This can be reproduced with the following:  

```python
In [15]: from extra_data import RunDirectory
    ...: with RunDirectory("/gpfs/exfel/exp/HED/202330/p900338/raw/r0052", "*S00000*") as
    ...:  fd:
    ...:     print(fd)
    ...: 
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-15-53c99aa5c4ec> in <module>
      1 from extra_data import RunDirectory
----> 2 with RunDirectory("/gpfs/exfel/exp/HED/202330/p900338/raw/r0052", "*S00000*") as fd:
      3     print(fd)
      4 

~/EXtra-data/extra_data/reader.py in __enter__(self)
    186     def __enter__(self):
    187         if not self.ctx_closes:
--> 188             raise Exception(
    189                 "Only DataCollection objects created by opening "
    190                 "files directly can be used in a 'with' statement, "

Exception: Only DataCollection objects created by opening files directly can be used in a 'with' statement, not those created by selecting from or merging others.

```  

## Possible fixes  

The current behaviour is not necessarily wrong, as it's difficult to know before hand if the `DataCollection` is going to be used as a context manager.

In the case of the correction, [we explicitly ignore the voview file](https://git.xfel.eu/detectors/pycalibration/-/merge_requests/793) as we want to simplify file access.  

I think it would be beneficial to clarify the error message and suggest to pass `_use_voview=False`.  

Another solution would be to [remove the check in `__enter__`](https://github.com/European-XFEL/EXtra-data/blob/d4c547c3af9a993ec2da41b5bf3d830dce9ab43e/extra_data/reader.py#L187) as the check whether to close files is done again  in [`__exit__`](https://github.com/European-XFEL/EXtra-data/blob/d4c547c3af9a993ec2da41b5bf3d830dce9ab43e/extra_data/reader.py#L198).  

I'm happy to contribute a solution, but would need guidance.

Thanks,
Cyril

**Comment by philsmt** [comment]
Thanks for spotting this.

I suspect it was unintentionally forgotten to pass `ctx_closes=True` when explicitly creating the `DataCollection` from the voview file. So another solution would be to allow this to happen. The intent of `with` is to ensure files are closed, and I suspect the data-backing files are closed as well when the file with virtual datasets is closed?

I'm sorry if I did not make this solution clear when discussing this.

**Comment by takluyver** [comment]
Yep, I'm pretty sure I just overlooked that it should be setting `ctx_closes=True`. I don't think there's any reason that shouldn't work with a virtual overview file.

>  The intent of with is to ensure files are closed

Yup, although as we've made file access more complicated for other reasons, closed files may be automatically reopened if you access something from them, so it's not a particularly strong guarantee.
